mlflow/server/js/src/components/ExperimentView.js

```diff
@@ -475,19 +478,46 @@ class ExperimentView extends Component {
       });
   }
 
-  onCheckbox(runUuid) {
-    const newState = Object.assign({}, this.state);
-    if (this.state.runsSelected[runUuid]) {
-      delete newState.runsSelected[runUuid];
-      this.setState(newState);
-    } else {
-      this.setState({
-        runsSelected: {
-          ...this.state.runsSelected,
-          [runUuid]: true,
+  /**
+   * Handler for a click event on a checkbox in the runs table. Handles both clicking individual
+   * runs and shift-clicking to select or deselect multiple contiguous runs.
+   * @param event Click event
+   * @param childrenIds Child run UUIDs of the current run
+   * @param index Index of the current run within sortedRunIds.
+   * @param sortedRunIds List of run UUIDs (both visible and collapsed) sorted by the current
+   *                     display order.
+   */
+  onCheckbox(event, childrenIds, index, sortedRunIds) {
+    const runUuid = sortedRunIds[index];
+    const minCheckboxIndex = this.state.lastCheckboxIndex !== undefined ?
+      Math.min(this.state.lastCheckboxIndex, index) : index;
+    const maxCheckboxIndex = this.state.lastCheckboxIndex !== undefined ?
+      Math.max(this.state.lastCheckboxIndex, index) + 1 : index + 1;
+    // Handle shift-clicks: Update selected state of all runs between the previously clicked and
+    // currently clicked run
+    const runsSelectedState = Object.assign({}, this.state.runsSelected);
+    if (event.shiftKey) {
+      _.range(minCheckboxIndex, maxCheckboxIndex).forEach(i => {
+        if (this.state.runsSelected[runUuid]) {
+          delete runsSelectedState[sortedRunIds[i]];
+        } else {
+          runsSelectedState[sortedRunIds[i]] = true;
         }
       });
     }
+    // If parent run is selected/unselected, select/unselect all child runs
```

Comment:
I have to wonder if this is the behavior people typically expect, or if they'd like a way to flip it on and off. Maybe @gioa would have opinions on how to deal with selecting child runs.

Reply:
Yeah people will definitely have use cases where they don't want this behavior. For example with hyperparam tuning, users might have metrics associated with the parent run (e.g. best-model loss) that they want to compare across parent runs - always selecting the children complicates this UX.

We could have a separate toggle (e.g. ctrl-click, which [is also a property](https://reactjs.org/docs/events.html#mouse-events) React passes on click events) for this behavior, and make the behavior either opt-in or opt-out based on whether the user is ctrl-clicking (it'd be more consistent with shift-clicking if the behavior were opt-in, but that'd also make it less discoverable).

Alternatively we could re-use the shift-key as a toggle (disable child-autoselection on shift-clicks) which is probably something we should do anyways, since it's weird IMO to shift-click to a parent run further down the page & have children runs even further down be selected - however this means there's no way to select multiple parents without selecting all the runs between them, which isn't great - we can ask Joy about these questions tomorrow.

Note: "the behavior" above = auto-[un]selecting child runs when the parent is [un]selected

Reply:
I made an assumption that the users always compare child runs. Did not know they also want to compare across parent runs.  What are the top most common comparisons? We should optimize for those.